### PR TITLE
lesson4

### PR DESCRIPTION
**Pull request by gavinzg**
- [x] This PR contains only information intended for public view

To view this PR's branch in binder:
```
https://mybinder.org/v2/gh/viettrung9012/python-codeacademy-sg/<branch>?urlpath=lab
```

### CODE REVIEW

File: Lesson4.ipynb [hunk 1]
```diff
@@ -29,6 +33,1310 @@
...
+    "team_no,team_name,team_captain,team_member,2018-04-02,2018-04-03\n",
+    "\n",
+    "1,TBD,FALSE,1-1,11980,10437\n",
+    "\n",
+    "1,TBD,FALSE,1-2,22935,13399\n",
+    "\n",
+    "However, csv file is a table, as we are familiar with Excel, to make it more clear, we can have a comparison with Excel we generally use in our daily work, the first row is the coloumn name. Start from the second row, they are the specific information of each person.\n",
+    "\n",
+    "You can open csv file by textEditer, you'll see the content same as above.\n",
+    "If you can open it with Excel, then it has the same lookings as .xml file."
```
Comment (puikwan) [inline Lesson4.ipynb]: Replace "You can open csv file by textEditer, you'll see the content same as above. If you can open it with Excel, then it has the same lookings as .xml file." with

If you open the csv file using a text Editor such as Notepad, you'll see the content as above.
If you open the csv file using Excel, the content will have the same format as a .xls file.

File: Lesson4.ipynb [hunk 2]
```diff
@@ -29,6 +33,1310 @@
...
+    "## Read csv file"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### What is csv file?\n",
+    "\n",
+    "A CSV is a comma separated values file which allows data to be saved in a table structured format. CSVs are a kind of spreadsheet same with Excel but with a .csv extension. Traditionally they take the form of a text file containing information separated by commas, hence the name.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Replace "CSVs are a kind of spreadsheet same with Excel but with a .csv extension." with
"A CSV file is similar to an Excel spreadsheet, though it doesn't have the style formatting and has a .csv extension instead."

File: Lesson4.ipynb [hunk 3]
```diff
@@ -29,6 +33,1310 @@
...
+    "\n",
+    "** Content in CSV Sample **\n",
+    "\n",
+    "The followings are the part of data from Biggest Loser:\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Replace "The followings are the part of data from Biggest Loser:"
with 
"The following sample data is from the Biggest Loser step challenge. For those unfamiliar with the Biggest Loser step challenge, this was a 4-week long challenge where teams (each consisting 5 members) combined their daily step count in order to reach these goals: 
1) to be the first team to reach 1 million steps
2) to have the highest overall step count amongst all teams
3) to show the most improved team step count (from weeks 1+2 to weeks 3+4)

File: Lesson4.ipynb [hunk 4]
```diff
@@ -29,6 +33,1310 @@
...
+    "\n",
+    "team_no,team_name,team_captain,team_member,2018-04-02,2018-04-03\n",
+    "\n",
+    "1,TBD,FALSE,1-1,11980,10437\n",
+    "\n",
+    "1,TBD,FALSE,1-2,22935,13399\n",
+    "\n",
+    "However, csv file is a table, as we are familiar with Excel, to make it more clear, we can have a comparison with Excel we generally use in our daily work, the first row is the coloumn name. Start from the second row, they are the specific information of each person.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Replace "However, csv file is a table, as we are familiar with Excel, to make it more clear, we can have a comparison with Excel we generally use in our daily work, the first row is the coloumn name. Start from the second row, they are the specific information of each person." with
"In the CSV format, the first header row consists of column names, while each subsequent row shows the data for each team member separated by a comma."

File: Lesson4.ipynb [hunk 5]
```diff
@@ -29,6 +33,1310 @@
...
+   "metadata": {},
+   "source": [
+    "\n",
+    "## How to read csv file in Python? ##\n",
+    "\n",
+    "To read and analyse the data, I recommend Pandas Library to implement the data analysis.\n",
+    "\n",
+    "### What is Pandas ###\n",
+    "\n",
+    "Pandas stands for “Python Data Analysis Library”. According to the Wikipedia page on Pandas, “the name is derived from the term “panel data”, an econometrics term for multidimensional structured data sets.” But I think it’s just a cute name to a super-useful Python library!\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: We discussed that pandas cannot be used in lesson 4, and is a bonus for lesson 8. Lesson 4 should be using normal python csv read.

File: Lesson4.ipynb [hunk 6]
```diff
@@ -6,7 +6,10 @@
    "source": [
     "## Lesson 4 Overview\n",
     "\n",
-    "### TODO"
+    " * Read csv file\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: We should be consistent, either **csv** or **CSV**, there are a lot of mixed here

File: Lesson4.ipynb [hunk 7]
```diff
@@ -29,6 +32,897 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "['team_no', 'team_name', 'team_captain', 'team_member', '2018-04-02', '2018-04-03', '2018-04-04', '2018-04-05', '2018-04-06', '2018-04-07', '2018-04-08', '2018-04-09', '2018-04-10', '2018-04-11', '2018-04-12', '2018-04-13', '2018-04-14', '2018-04-15', '2018-04-16', '2018-04-17', '2018-04-18', '2018-04-19', '2018-04-20', '2018-04-21', '2018-04-22', '2018-04-23', '2018-04-24', '2018-04-25', '2018-04-26', '2018-04-27', '2018-04-28', '2018-04-29']\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: You can do a clear all output before saving the file, so we can ask everyone to try to run the code

File: Lesson4.ipynb (hunk 6 above)
Comment (viettrung9012) [reply Lesson4.ipynb]: I think we can have **CSV file**, **Python csv library**, and **.csv extension**

File: Lesson4.ipynb [hunk 8]
```diff
@@ -29,6 +32,897 @@
...
+      "['14', '', 'FALSE', '14-3', '11328', '4887', '5037', '13627', '6243', '14736', '5100', '11238', '4886', '3964', '5764', '10180', '3259', '6302', '4537', '7275', '78', '2969', '1989', '11312', '5910', '', '', '', '', '', '', '']\n",
+      "['14', '', 'FALSE', '14-4', '942', '7664', '2600', '10054', '', '7397', '', '', '', '15585', '15000', '11000', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '']\n",
+      "['14', '', 'FALSE', '14-5', '3175', '3853', '4174', '6633', '1553', '1675', '3265', '4693', '2987', '4383', '2640', '922', '2791', '4261', '4999', '3032', '4678', '3701', '5082', '5665', '6311', '3983', '2954', '3960', '3343', '3862', '6578', '3139']\n"
+     ]
+    }
+   ],
+   "source": [
+    "### Print CSV file\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: what does **with** do? Maybe a short description about the syntax so we don't forget during lesson

File: Lesson4.ipynb [hunk 9]
```diff
@@ -29,6 +32,897 @@
...
+    "    readCSV = csv.reader(csvfile)\n",
+    "    for row in readCSV:\n",
+    "        print(row[0],row[1],row[2])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before we do the next, let me introduce the enumerate() first.\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: enumerate() **function**

File: Lesson4.ipynb [hunk 10]
```diff
@@ -29,6 +32,897 @@
...
+    "\n",
+    "Enumerate is a built-in function of Python. It allows us to loop over something and have an automatic counter. (More details about enumerate you can find here: http://book.pythontips.com/en/latest/enumerate.html)\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: **e**numerate

File: Lesson4.ipynb [hunk 11]
```diff
@@ -29,6 +32,897 @@
...
+    "\n",
+    "### Print specifical line\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: Print a specific line

File: Lesson4.ipynb [hunk 12]
```diff
@@ -29,6 +32,897 @@
...
+    "            print(row1)\n",
+    "            break\n",
+    "        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we mentioned from the last week, the index of a List starts from 0, this rule suits for everywhere in coding. \n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: can use "before" or "lesson 2" instead of "the last week", because I think we introduced index 0 at lesson 2

File: Lesson4.ipynb [hunk 13]
```diff
@@ -29,6 +32,897 @@
...
+      "['1', 'TBD', 'FALSE', '1-1', '11980', '10437', '17616', '24586', '16136', '13700', '39812', '9195', '12855', '11309', '23606', '11848', '6120', '6254', '8754', '6469', '8849', '9911', '7709', '534', '13465', '7341', '11230', '7878', '11029', '8790', '9006', '21942']\n"
+     ]
+    }
+   ],
+   "source": [
+    "### Print the data of first line\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.reader(csvfile)\n",
+    "    for i, row in enumerate(readCSV):\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: can we use **rowIndex, row** instead of **i, row**

File: Lesson4.ipynb [hunk 14]
```diff
@@ -29,6 +32,897 @@
...
+    "        if i == 1:\n",
+    "            row1 = row\n",
+    "            print(row1)\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: why don't we just use print(row)?

File: Lesson4.ipynb [hunk 15]
```diff
@@ -29,6 +32,897 @@
...
+    "\n",
+    "*** Steps statistics ***\n",
+    "\n",
+    "In this part, the premisses we need to know are two things.\n",
+    "\n",
+    "   * Which columns are the data for steps. \n",
+    "   * Which rows are the data for attendees.\n",
+    "\n",
+    "Seen from the table, we can easily figure out the answers.\n",
+    "   * -- Columns start from the fourth column to the end, that is to say, the index satrts from 3.\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: no, the index starts from 4, and **starts** not **satrts**

File: Lesson4.ipynb [hunk 16]
```diff
@@ -29,6 +32,897 @@
...
+    "   * -- Rows start from the second row to the end, that is to say, the index satrts from 1 ( We also can say greater than 0, not including 0 ).\n"
```
Comment (viettrung9012) [inline Lesson4.ipynb]: same here, **starts** typo

File: Lesson4.ipynb (hunk 10 above)
Comment (viettrung9012) [reply Lesson4.ipynb]: it's probably better if you give a quick overview of enumerate compared to normal list "for" loop:
- "for" with list, only item, no index
- "for" with enumerate(list), can have item and index

File: Lesson4.ipynb [hunk 17]
```diff
@@ -29,6 +32,897 @@
...
+      "Total steps of Member70 is:108292\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Print total steps per person\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:                          # Open the csv file by path\n",
+    "    readCSV = csv.reader(csvfile)                                        # Read the csv file\n",
+    "    for i,rows in enumerate(readCSV):                                    # Loop statement to fetch the data                     \n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: let's just use "rowIndex, row" here

File: Lesson4.ipynb [hunk 18]
```diff
@@ -29,6 +32,897 @@
...
+    "        if i > 0:                                                        # Rows start from 1\n",
+    "            row = rows                                                   # Define row for each line\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: just use row in the loop start line, don't reassign

File: Lesson4.ipynb [hunk 19]
```diff
@@ -29,6 +32,897 @@
...
+    "            totalSteps = 0                                               # Define a parameter to store each attendees' total steps\n",
+    "            for j, step in enumerate(row):                               # Caculate the steps starts by using Loop statement\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: Calculate

File: Lesson4.ipynb [hunk 20]
```diff
@@ -29,6 +32,897 @@
...
+    "                if j > 3 and step != '':                                 # Starts from cloumn 4\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: column

File: Lesson4.ipynb [hunk 21]
```diff
@@ -29,6 +32,897 @@
...
+    "                    totalSteps = totalSteps + int(step)                 \n",
+    "            print(\"Total steps of Member\" + str(i) + \" is:\" + str(totalSteps)) # After \"for\" statement, then get the total steps"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "From the code above, we get total steps for every attendees, and print them one by one.\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: attendee/participant in singular

File: Lesson4.ipynb [hunk 22]
```diff
@@ -29,6 +32,897 @@
...
+    "\n",
+    "I think somebody have noticed a something, when caculating the total steps, we use a new thing in the code on Line 11\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: calculating

File: Lesson4.ipynb [hunk 23]
```diff
@@ -29,6 +32,897 @@
...
+    "\n",
+    "Let's check the code block again.\n",
+    "```python \n",
+    "      totalSteps = 0\n",
+    "       for j, step in enumerate(row):  \n",
+    "           if j > 3 and step != '':                                \n",
+    "               totalSteps = totalSteps + int(step) \n",
+    "```\n",
+    "\n",
+    "We define totalSteps and set initialization to 0, which is an integer, meanwhile, step in row is the data from table, which is string, we need to convert the string to integer before plus.\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: before adding

File: Lesson4.ipynb [hunk 24]
```diff
@@ -29,6 +32,897 @@
...
+    "with open('Biggest Loser 2018.csv') as csvfile:                          # Open the csv file by path\n",
+    "    readCSV = csv.reader(csvfile)                                        # Read the csv file\n",
+    "    for i,rows in enumerate(readCSV):                                    # Loop statement to fetch the data                     \n",
+    "        if i > 0:                                                        # Rows start from 1\n",
+    "            row = rows                                                   # Define row for each line\n",
+    "            totalSteps = 0                                               # Define a parameter to store each attendees' total steps\n",
+    "            for j, step in enumerate(row):                               # Caculate the steps starts by using Loop statement\n",
+    "                if j > 3 and step != '':                                 # Starts from cloumn 4\n",
+    "                    totalSteps = totalSteps + int(step)                 \n",
+    "            print(\"Total steps of Member\" + str(i) + \" is:\" + str(totalSteps)) # After \"for\" statement, then get the total steps"
```
Comment (viettrung9012) [inline Lesson4.ipynb]: a quick reminder about indentation here could be useful, why print is indented outside of the second for loop but inside the first for loop

File: Lesson4.ipynb [hunk 25]
```diff
@@ -29,6 +32,897 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Redo Lesson 3 Questions\n",
+    "\n",
+    "#### Print names of people with steps count more than 100k/300k/400k\n",
+    "\n",
+    "In lesson 3, we have already known how to print the names who clock the steps more than 100k, but the data is in the List  prepared before computing. Now let's use the data from CSV file to answer the same questions with last class.\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: the names of people

Comment (viettrung9012) [reply Lesson4.ipynb]: same questions from last lesson

File: Lesson4.ipynb [hunk 26]
```diff
@@ -29,6 +32,897 @@
...
+    "We define totalSteps and set initialization to 0, which is an integer, meanwhile, step in row is the data from table, which is string, we need to convert the string to integer before plus.\n",
+    "\n",
+    "As a result, int() is used for converting other types to integer. We have to do it when in computing."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Redo Lesson 3 Questions\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: apparently I don't think we did these in lesson 3

File: Lesson4.ipynb (hunk 19 above)
Comment (viettrung9012) [reply Lesson4.ipynb]: maybe stepIndex, step is nicer compared to j, step

File: Lesson4.ipynb [hunk 27]
```diff
@@ -29,6 +32,897 @@
...
+    }
+   ],
+   "source": [
+    "#Print Data by column index and column name\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.DictReader(csvfile)\n",
+    "    header = readCSV.fieldnames\n",
+    "    for i, rows in enumerate(readCSV):\n",
+    "        print(rows[header[0]], rows[header[1]], rows['team_captain'], rows['team_member'])"
```
Comment (viettrung9012) [inline Lesson4.ipynb]: This code is kind of hard to read, maybe some explanation or variable assignment could make it nicer
```Python
team_number = header[0]
team_name = header[1]
team_captain = header[2]
team_member = header[3]
print(rows[team_number], rows[team_name], rows[team_captain], rows[team_member]
```

File: Lesson4.ipynb [hunk 28]
```diff
@@ -29,6 +32,897 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "#Print Data by column index and column name\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.DictReader(csvfile)\n",
+    "    header = readCSV.fieldnames\n",
+    "    for i, rows in enumerate(readCSV):\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: maybe rowIndex, row for clarity

File: Lesson4.ipynb [hunk 29]
```diff
@@ -29,6 +32,897 @@
...
+   "source": [
+    "# Print total steps per person by using DictReader\n",
+    "import csv\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.DictReader(csvfile)\n",
+    "    header = readCSV.fieldnames\n",
+    "    for i,rows in enumerate(readCSV):\n",
+    "        totalSteps = 0\n",
+    "        for j, dateHeader in enumerate(header):\n",
+    "            if j > 3 and rows[header[j]] != '':\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: maybe assign a variable to header[j] could make the code easier to read

File: Lesson4.ipynb [hunk 30]
```diff
@@ -29,6 +32,897 @@
...
+    "        return totalSteps\n",
+    "\n",
+    "    NamesForStepsMoreThan100K = ''\n",
+    "    NamesForStepsMoreThan300K = ''\n",
+    "    NamesForStepsMoreThan400K = ''\n",
+    "    \n",
+    "    for i,rows in enumerate(readCSV):\n",
+    "        totalSteps = sumStepsPerPerson()\n",
+    "        if totalSteps >= 100000:\n",
+    "            NamesForStepsMoreThan100K = NamesForStepsMoreThan100K + rows['team_member'] + ' '\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: let's do one only, 300k, 3 different conditions are hard to follow. And we should explain on String concatenation and variable re-assignment here too

File: Lesson4.ipynb (hunk 23 above)
Comment (viettrung9012) [reply Lesson4.ipynb]: this sentence is too long, break it down so maybe it is easier to read

File: Lesson4.ipynb [hunk 31]
```diff
@@ -29,6 +32,897 @@
...
+    "Last week, we know str(), which used for converting other types to string. Now I'm going to talk about int().\n",
+    "\n",
+    "When fetching the data from a table, the data types are the same -- All are string type.\n",
+    "\n",
+    "Let's check the code block again.\n",
+    "```python \n",
+    "      totalSteps = 0\n",
+    "       for j, step in enumerate(row):  \n",
+    "           if j > 3 and step != '':                                \n",
+    "               totalSteps = totalSteps + int(step) \n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: a quick explanation about variable re-assignment here could be useful

File: Lesson4.ipynb [hunk 32]
```diff
@@ -29,6 +32,897 @@
...
+    "    print(\"Steps more than 100K:\" + NamesForStepsMoreThan100K + \"/n\")\n",
+    "    print(\"Steps more than 300K:\" + NamesForStepsMoreThan300K + \"/n\")\n",
+    "    print(\"Steps more than 400K:\" + NamesForStepsMoreThan400K + \"/n\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## PrintStepsByName ##"
```
Comment (viettrung9012) [inline Lesson4.ipynb]: Print total steps for each team

File: Lesson4.ipynb [hunk 33]
```diff
@@ -29,6 +32,897 @@
...
+    "def sumSteps(header):\n",
+    "    totalSteps = 0\n",
+    "    for i, dateHeader in enumerate(header):\n",
+    "            if i > 3 and rows[header[i]] != '':\n",
+    "                totalSteps = totalSteps + int(rows[header[i]])\n",
+    "    return totalSteps\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.DictReader(csvfile)\n",
+    "    header = readCSV.fieldnames\n",
+    "    stepsOfTeam1 = 0\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: wouldn't another function here/double for loop works better than a plethora of if/elif and steps total?

File: Lesson4.ipynb [hunk 34]
```diff
@@ -29,6 +32,963 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### What is CSV file?\n",
+    "\n",
+    "A CSV is a comma separated values file which allows data to be saved in a table structured format. A CSV file is similar to an Excel spreadsheet, though it doesn't have the style formatting and has a .csv extension instead. Traditionally they take the form of a text file containing information separated by commas, hence the name.\n",
+    "\n",
+    "** Content in CSV Sample **\n",
+    "\n",
+    "Start from this lesson, we will use the data from Biggest Loser. For those unfamiliar with the Biggest Loser step challenge, this was a 4-week long challenge where teams (each consisting 5 members) combined their daily step count in order to reach these goals:\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "Start from this lesson" should read "Starting from this lesson"

File: Lesson4.ipynb [hunk 35]
```diff
@@ -29,6 +32,963 @@
...
+    "\n",
+    "To read and analyse the data, we use the Python csv library. CSV literally stands for comma separated variable, where the comma is what is known as a \"delimiter.\"\n",
+    "\n",
+    "    Tips: Whatever the library you need, you should import all the necessary libraries at the beginnng of the code.\n",
+    "    \n",
+    "## Let's start\n",
+    "\n",
+    "### Print the csv file\n",
+    "\n",
+    "    1. The first thing we need to do is importing CSV, then we can use the functions under Python csv library.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "The first thing we need to do is importing CSV, then we can use the functions under Python csv library." should read "The first thing we need to do is import the Python CSV library, so we can use the functions in that library."

File: Lesson4.ipynb [hunk 36]
```diff
@@ -29,6 +32,963 @@
...
+    "    2. Open the file by path\n",
+    "    3. By reading CSV file, we can use reader function under Python csv library\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "By reading CSV file," should read "To read the CSV file,"

File: Lesson4.ipynb [hunk 37]
```diff
@@ -6,7 +6,10 @@
    "source": [
     "## Lesson 4 Overview\n",
     "\n",
-    "### TODO"
+    " * Read CSV file\n",
+    " * List contents of the CSV file\n",
+    " * Redo lesson 3 questions after reading CSV\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Please remove the point on "Redo lesson 3 questions after reading CSV". Don't believe this was covered in Lesson 4.

File: Lesson4.ipynb [hunk 38]
```diff
@@ -29,6 +32,655 @@
...
+    "            print(\"Total steps of Member\" + str(row_index) + \" is:\" + str(totalSteps)) # After \"for\" statement completes running, totalSteps is the answer we want"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "From the code block above, we get total steps for every attendee, and print them one by one.\n",
+    "\n",
+    "I think somebody have noticed something, when calculating the total steps, we use a new thing in the code on Line 10\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: ""I think somebody have noticed something" should read "You may have noticed"

File: Lesson4.ipynb [hunk 39]
```diff
@@ -29,6 +32,655 @@
...
+   "source": [
+    "### Total Steps for each member\n",
+    "\n",
+    "After the Biggest Loser event, I think all the attendees wonder how many steps they clock in total.\n",
+    "\n",
+    "It's not a hard question, let's do coding to get the answer.\n",
+    "\n",
+    "*** Steps statistics ***\n",
+    "\n",
+    "In this part, the premisses we need to know are two things.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "In this part, the premisses we need to know are two things." should read "In this part, we should know two things."

File: Lesson4.ipynb [hunk 40]
```diff
@@ -29,6 +32,655 @@
...
+    "From the code block above, we get total steps for every attendee, and print them one by one.\n",
+    "\n",
+    "I think somebody have noticed something, when calculating the total steps, we use a new thing in the code on Line 10\n",
+    "\n",
+    "```python \n",
+    "int(step)\n",
+    "```\n",
+    "** **\n",
+    "\n",
+    "Last week, we know str(), which used for converting other types to string. Now I'm going to talk about int().\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "Last week, we know str(), which used for converting other types to string." should read "Last week, we learnt str(), which is used for converting other types to string."

File: Lesson4.ipynb [hunk 41]
```diff
@@ -29,6 +32,655 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Redo Lesson 3 Questions\n",
+    "\n",
+    "#### Print names of people with steps count more than 100k/300k/400k\n",
+    "\n",
+    "In lesson 3, we have already known how to print the names of people who clock the steps more than 100k, but the data is in the List which prepared before computing. Now let's use the data from a CSV file to answer the same questions from last lesson.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Remove reference to lesson 3 since this was an optional challenge question (as discussed with Trung)

File: Lesson4.ipynb [hunk 42]
```diff
@@ -29,6 +32,655 @@
...
+    "    print(header)\n",
+    "    "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "From the results printed above, the \"header\" is a list consisting the column names of the table.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "a list consisting the column names" should read "a list of the column names"

File: Lesson4.ipynb [hunk 43]
```diff
@@ -29,6 +32,655 @@
...
+    "    for header_index, fieldNames in enumerate(header):                   # looping header for counting the steps\n",
+    "        steps_in_column = row[header[header_index]]\n",
+    "        if header_index > 3 and steps_in_column != '':                   # Step amount starts from column 4\n",
+    "            totalSteps = totalSteps + int(steps_in_column)\n",
+    "    return totalSteps\n",
+    "```\n",
+    "** **\n",
+    "By reusing the function, we can get the total steps per person, then compare with 100k/300k/400k\n",
+    "\n",
+    "We can define three string paramters, **NamesForStepsMoreThan100K**, **NamesForStepsMoreThan300K**, **NamesForStepsMoreThan400K** to store the names which satisfy the conditions.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "string paramters" should read "string parameters"

File: Lesson4.ipynb [hunk 44]
```diff
@@ -29,6 +32,655 @@
...
+    "        if totalSteps >= 100000:\n",
+    "            NamesForStepsMoreThan100K = NamesForStepsMoreThan100K + row['team_member'] + ' '\n",
+    "    print(\"Steps more than 100K:\" + NamesForStepsMoreThan100K)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Follow the code block above, can you complete a code block to print the names of people with steps count more than 300k and 400k**"
```
Comment (puikwan) [inline Lesson4.ipynb]: "Follow the code block above" should read "Following the code block above"

File: Lesson4.ipynb [hunk 45]
```diff
@@ -29,6 +32,655 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Print total steps for each team\n",
+    "\n",
+    "From the code above, we complete a function for counting total steps for each attendee.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "From the code above, we complete a function" should read "From the code above, we have completed a function"

File: Lesson4.ipynb [hunk 46]
```diff
@@ -29,6 +32,655 @@
...
+    "    for row in readCSV:\n",
+    "        if row['team_name'] == team_name:        # If the \"team_name\" in a row equals the target team_name, then plus it\n",
+    "            teamTotalSteps = teamTotalSteps + totalStepsByRow(row, header)\n",
+    "```\n",
+    "** **\n",
+    "### Loop for team names\n",
+    "\n",
+    "To print the total steps for each team, let's do a loop for the team names outside the code block above.\n",
+    "\n",
+    "Please refer the code block below:\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "Please refer the code block below" should read ""Please refer to the code block below"

File: Lesson4.ipynb [hunk 47]
```diff
@@ -29,6 +32,655 @@
...
+    "for name in team_names:\n",
+    "    totalStepsByTeam = getTotalStepsByTeamNameFromCSV(name)\n",
+    "    print(\"Total steps of Team \" + name + \" is: \" + str(totalStepsByTeam))\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "That's it for Lesson 2!\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "That's it for Lesson 2!" should read "That's it for Lesson 4!"

File: Lesson4.ipynb [hunk 48]
```diff
@@ -29,6 +32,655 @@
...
+    "    with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "        readCSV = csv.DictReader(csvfile)\n",
+    "        team_names = []\n",
+    "        for row in readCSV:\n",
+    "            if row['team_name'] not in team_names:\n",
+    "                team_names.append(row['team_name'])\n",
+    "    return team_names\n",
+    "\n",
+    "def getTotalStepsByTeamNameFromCSV(team_name):\n",
+    "    with open('Biggest Loser 2018.csv') as csvfile:\n",
```
Comment (viettrung9012) [inline Lesson4.ipynb]: either double open file or we can use list casting:
```Python
# Print total steps by each team
import csv

def totalStepsByRow(row, header):
    totalSteps = 0
    for header_index, fieldNames in enumerate(header):
        steps_in_column = row[header[header_index]]
        if header_index > 3 and steps_in_column != '':
            totalSteps = totalSteps + int(steps_in_column)
    return totalSteps

def get_team_names(steps_data):
    team_names = []
    for row in steps_data:
        if row['team_name'] not in team_names:
            team_names.append(row['team_name'])
    return team_names

def get_total_steps(team_name, steps_data, header):
    total_steps = 0
    for row in steps_data:
        if row['team_name'] == team_name:
            total_steps = total_steps + totalStepsByRow(row, header)
    return total_steps

with open('Biggest Loser 2018.csv') as csvfile:
    csvdata = csv.DictReader(csvfile)                           # Returns an iterator object, and iterator can only loop once!
    header = csvdata.fieldnames                                 # Get the header
    steps_data = list(csvdata)                                  # Convert the steps data to list since we loop over more than once
    team_names = get_team_names(steps_data)
    for name in team_names:
        team_total_steps = get_total_steps(name, steps_data, header)
        print('Total steps of Team ' + name + ' is: ' + str(team_total_steps))
```

Comment (viettrung9012) [reply Lesson4.ipynb]: Can add an open-ended question as a challenge like: Currently our simple implementation loops through the data multiple times (18 teams, 1 for getting team name and once for each team). This is not very efficient because we keep revisits the same data. Imagine we have more teams the number of steps will keep growing. Can we reduce the number of loops?

File: Lesson4.ipynb (hunk 8 above)
Comment (puikwan) [reply Lesson4.ipynb]: Answered below

File: Lesson4.ipynb [hunk 49]
```diff
@@ -29,6 +33,737 @@
...
+    "\n",
+    "*** The following sample data is from the Biggest Loser step challenge. ***\n",
+    "\n",
+    "     team_no,team_name,team_captain,team_member,2018-04-02,2018-04-03\n",
+    "\n",
+    "     1,TBD,FALSE,1-1,11980,10437\n",
+    "\n",
+    "     1,TBD,FALSE,1-2,22935,13399\n",
+    "\n",
+    "However, CSV file is a table, as we are familiar with Excel, to make it more clear, we can have a comparison with Excel we generally use in our daily work. The first row is the coloumn name, start from the second row, they are the specific information of each person.\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "coloumn" should read "column"

File: Lesson4.ipynb [hunk 50]
```diff
@@ -29,6 +33,737 @@
...
+    "            "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Write a function for counting total steps\n",
+    "\n",
+    "Steps start from column 5, we can count the total steps by looping header index starts from 4, which is column_index > 3\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: "we can count the total steps by looping header index starts from 4" should read "we can count the total steps by looping header index starting from 4"

File: Lesson4.ipynb [hunk 51]
```diff
@@ -29,6 +33,737 @@
...
+    "\n",
+    "def getTotalStepsByTeamNameFromData(stepsData, header, team_name):\n",
+    "    teamTotalSteps = 0\n",
+    "    for row in stepsData:\n",
+    "        if row['team_name'] == team_name:\n",
+    "            teamTotalSteps = teamTotalSteps + totalStepsByRow(row, header)\n",
+    "    return teamTotalSteps\n",
+    "\n",
+    "with open('Biggest Loser 2018.csv') as csvfile:\n",
+    "    readCSV = csv.DictReader(csvfile)           # Returns an iterator object, and can only loop once!\n",
```
Comment (puikwan) [inline Lesson4.ipynb]: Remove this comment
